Add functions to support Standard Swizzle textures

isplunke opened the pull request:
Implement z-order curve "swizzling" for 2D and 3D textures to support the pixel order used for "Standard Swizzle" as defined by `D3D12_TEXTURE_LAYOUT_64KB_STANDARD_SWIZZLE`.

> These functions only perform the row-major <=> z-order curve operation. 64KB alignment, miptail packing, and other requirements are handled in other distinct operations.

isplunke committed:
[Draft]
- needs to be tested.  where to get dx12 boilerplate?  test just that texute still looks correct, or also test performance improves when roteated?
- maybe merge functions?  would then pass a boolean/enum to determine rowToSwizzle versus swizzleToRow

Added functions that converts pixel order from row major to standard swizzle and from standard swizzle to row major.

Functions for both 1 Image, and an array of Images.

Folwoing standard DirectXTex pattern: Uses Image and TexMetadata as input.  Outputs/initializes a ScratchImage.

isplunke commented:
@microsoft-github-policy-service agree company="Microsoft"

walbourn commented on DirectXTex/DirectXTexImage.cpp:
```diff
@@ -567,6 +567,78 @@ HRESULT ScratchImage::InitializeFromImage(const Image& srcImage, bool allow1D, C
+
+    size_t bytesPerPixel = BitsPerPixel(srcImage.format) / 8;
+
+    uint32_t xBytesMask = 0b1010101010101010;
```
I believe the 3D case is to process the entire 'slice' at a time since they are continuous in memory and you use:

```
size_t swizzleIndex = _pdev_u32(x, maskx) + _pdep_u32(y, masky) + _pdep_u32(z, maskz);
```

isplunke replied:
Regarding bit pattern versus bytes per pixel:

I also was confused by this.  The variable bit patterns are to be used with address offset.  Notice how they ignore the lower bits for alignment?

I'm using a bit pattern for index.  And then converting to bpp specific memory offset later.

walbourn replied:
Note I had to edit the 3D patterns. My source encoded them slightly shifted for a LUT thing which doesn't apply to our usage.

walbourn replied:
There are different patterns based on the BPP. The index is the index to the pixel regardless of its size, not the byte.

Note that this means the 'pitch' is ignored completely in the swizzled image and you can end up with some extra memory at the end for cases where pitch is not the same as ``bpp*w``. This space is zero filled for most cases, although apparently for BC they replicate the last block data instead.

isplunke replied:
"The index is the index to the pixel regardless of its size"
Yes, that's why I'm swizzleing the index.  Because there's only one pattern.

"the 'pitch' is ignored completely in the swizzled image"
Yes, that's why .rowPitch is only used in rowMajorOffset.  Not is Swizzle[Index|Offset].

I don't see the problem.

walbourn commented on DirectXTex/DirectXTexImage.cpp:
```diff
@@ -567,6 +567,78 @@ HRESULT ScratchImage::InitializeFromImage(const Image& srcImage, bool allow1D, C
+        {
+            uint32_t swizzleIndex = _pdep_u32(x, xBytesMask) + _pdep_u32(y, yBytesMask);
+
+            uint8_t* sourcePixelPointer = sptr + y * srcImage.rowPitch + x * bytesPerPixel;
```
Note that ``sourcePixelPointer`` can and should be ``const``

walbourn commented:
Note for testing, I'll add coverage to the [test suite](https://github.com/walbourn/directxtextest) which will likely just do a swizzle and deswizzle and make sure the bits match.

As for determining if the swizzle is correct, I probably need to make a test app that renders using WARP which supports standard swizzle.

isplunke committed:
Moved functionst to new file DirectTexSwizzle
Added new file to .vcxproj, other.vcxproj, and Cmake

Merged to/from functions togeather
memcpy src is const
added non-AVX2 deposit_bits

isplunke committed:
cleanup move

isplunke commented on DirectXTex/DirectXTexImage.cpp:
```diff
@@ -627,6 +699,84 @@ HRESULT ScratchImage::InitializeArrayFromImages(const Image* images, size_t nIma
+        {
+            for (size_t x = 0; x < srcImage[imageIndex].width; x++)
+            {
+                uint32_t swizzleIndex = _pdep_u32(x, xBytesMask) + _pdep_u32(y, yBytesMask);
```
after merging the to/from swizzle functions.  I ended up using a ternary to switch offsets, instead of _pext_u32.

walbourn replied:
OK. It is probably going to be more efficient to do a different loop for each case in terms of speed since this can potentially be a lot of pixels.

walbourn replied:
IOW:, I expect each of these three functions to have four cases:

```
if (toSwizzle)
{
    // Convert rowmajor linear to a z-order curve
    if (IsCompressed(format))
    {
        size_t blockBytes = (format == DXGI_FORMAT_BC1 || format == DXGI_FORMAT_BC4) ? 8 : 16;

        size_t nbw = std::max<size_t>(1, (width + 3) / 4); 
        size_t nbh = std::max<size_t>(1, (height + 3) / 4);
        for(size_t y = 0; y < nbh; ++y)
        {
            for(size_t x = 0; x < nbw; ++x)
            {
                // Swizzle index using pdep from x/y using blockBytes as each 'pixel' size
            }
        }
    }
    else
    {
        size_t pixelBytes = BitsPerPixel(format) / 8;
        for(size_t y = 0; y < height; ++y)
        {
            for(size_t x = 0; x < width; ++x)
            {
                // Swizzle index using pdep from x/y using pixelBytes as each 'pixel' size
            }
        }
    }
}
else
{
    // Convert a z-order curve to rowmajor linear.
    if (IsCompressed(format))
    {
        size_t blockBytes = (format == DXGI_FORMAT_BC1 || format == DXGI_FORMAT_BC4) ? 8 : 16;

        size_t nbw = std::max<size_t>(1, (width + 3) / 4); 
        size_t nbh = std::max<size_t>(1, (height + 3) / 4);
        for(size_t y = 0; y < nbh; ++y)
        {
            for(size_t x = 0; x < nbw; ++x)
            {
                // Destination index using pext from x/y using blockBytes as each 'pixel' size
            }
        }
    }
    else
    {
        size_t pixelBytes = BitsPerPixel(format) / 8;
        for(size_t y = 0; y < height; ++y)
        {
            for(size_t x = 0; x < width; ++x)
            {
                // Destination pixel index using pext from x/y using pixelBytes as each 'pixel' size
            }
        }
    }
}

isplunke replied:
Nit: "The current approach you are using means that BOTH source and dest will be accessed 'randomly' "
Only one would have been accessed "randomly"(swizzled), the other was rowMajor and thus linearly.

But I still refactored it to always read rowMajor|linerar, and only write "randomly"(swizzle).

walbourn replied:
I agree that the compressed and non-compressed cases should be able to be combined.

isplunke commented on DirectXTex/DirectXTexImage.cpp:
```diff
@@ -567,6 +567,78 @@ HRESULT ScratchImage::InitializeFromImage(const Image& srcImage, bool allow1D, C
+    {
+        for (size_t x = 0; x < srcImage.width; x++)
+        {
+            uint32_t swizzleIndex = _pdep_u32(x, xBytesMask) + _pdep_u32(y, yBytesMask);
```
Odd,
When I've used sse4 instructions previously, gcc warned that not all x64 cpus had it.  I'm surprised that Visual Studio, with all of it's bells and whistles, didn't catch me on this.

walbourn replied:
VS assumes if you are using intrinsics, you are on your own :)  Clang will complain if you use an intrinsic if you don't have the appropriate  -march` set.

isplunke committed:
added 3D

isplunke commented on DirectXTex/DirectXTexImage.cpp:
```diff
@@ -567,6 +567,78 @@ HRESULT ScratchImage::InitializeFromImage(const Image& srcImage, bool allow1D, C
+    if (!dptr)
+        return E_POINTER;
+
+    size_t bytesPerPixel = BitsPerPixel(srcImage.format) / 8;
```
I've been reading up on this.  https://learn.microsoft.com/en-us/windows/win32/direct3d10/d3d10-graphics-programming-guide-resources-block-compression

Block Compression works on 4x4 blocks of pixels.  Since pixels are smushed together, we really can't swizzle them in this format.  The only solution I thought of is to Decompress them, [un]swizzle them, and re-Compress them.

However! our Compress function needs TEX_COMPRESS_FLAGS and a threshold.  Since the source image doesn't store them, I'm not sure where to get good values.  Any Ideas?

walbourn replied:
You can just treat each block as if it were an 8byte or 16byte pixel. That's what Standard Swizzle is doing for BC.

walbourn replied:
To clarify the Compressed case is done by treating it as a 1/4th image where each block is a "pixel".

```
size_t blockBytes = (format == DXGI_FORMAT_BC1 || format == DXGI_FORMAT_BC4) ? 8 : 16;
size_t nbw = std::max<size_t>(1, (width + 3) / 4); 
size_t nbh = std::max<size_t>(1, (height + 3) / 4);
for(size_t y = 0; y < nbh; ++y)
{
    for(size_t x = 0; x < nbw; ++x)
    {

    }
}
```

walbourn replied:
I really need to add a helper for getting the 8 or 16 value from a format for BC. I'll fix that shortly.

isplunke replied:
Isn't that what I was doing originally?

walbourn replied:
If you merge the latest main, there is a new ``BytesPerBlock`` helper.

isplunke committed:
Almost handel Block Compression.
TODO what flags/threshold to use to re-Compress?

isplunke committed:
Fix 3d swizzle idexing into multiple images

isplunke commented on DirectXTex/DirectXTexImage.cpp:
```diff
@@ -627,6 +699,84 @@ HRESULT ScratchImage::InitializeArrayFromImages(const Image* images, size_t nIma
+_Use_decl_annotations_
+HRESULT ScratchImage::InitializeImagesFromSwizzles(const Image* srcImage, size_t nImages, CP_FLAGS flags) noexcept
+{
+    HRESULT hr = Initialize2D(srcImage[0].format, srcImage[0].width, srcImage[0].height, nImages, 1, flags);
```
Looked at  TransformImage in DirectXTexMisc.cpp.  I only saw mips being used in the 3d case.  And even in that one, we're kind of just iterating for index = 0; index < nImages.

In the case: array of 2d images, we're iterating through each image independantley.  So I don't think there's anything new to do there?

In the 3d case.  I should separate slice (which is being swizzled) and layer( which is not being swizzled).   But what order is the array of srcImages?  layerMajor or sliceMajor?

walbourn replied:
For the 3D case, my understanding is you would do:

```
size_t width = mdata.width;
size_t height = mdata.height;
size_t depth = mdata.depth;
for (size_t level = 0; level < mdata.mipLevels; ++level)
{
    auto destPtrBase = result.GetImage(level, 0, 0); // All slices of a given depth are always continuous in ScratchImage.
    for(size_t z = 0; z< depth; ++z)
    {
        auto srcImage = result.GetImage(level, 0, z);
        for(size_t h = 0; h < height; ++h)
        {
            for(size_t x = 0; 0 < width; ++w)
            {
                // swizzle using x, y, z copying from srcImage to destPtrBase + swizzleIndex
            }
        }
    }

    if (height > 1)
        height >>= 1;

    if (width > 1)
        width >>= 1;

    if (depth > 1)
        depth >>= 1;
}
```

walbourn commented on DirectXTex/DirectXTexSwizzle.cpp:
```diff
@@ -0,0 +1,303 @@
+//-------------------------------------------------------------------------------------
+// DirectXTexImage.cpp
```
// DirectXTexSwizzle.cpp

walbourn commented on DirectXTex/DirectXTexSwizzle.cpp:
```diff
@@ -0,0 +1,303 @@
+    if (IsCompressed(srcImage.format))
+    {
+        wide = ScratchImage();
+        HRESULT de = Decompress(srcImage, srcImage.format, wide);
```
Since Block Compression is lossy, decompress/swizzle/compress is not a high-quality solution. You can just swizzle it directly treating each block as if it were an 8byte or 16byte pixel. Same algorithm.

isplunke replied:
Duplicate of https://github.com/microsoft/DirectXTex/pull/515#discussion_r1765398888

walbourn commented on DirectXTex/DirectXTex_Desktop_2022.vcxproj:
```diff
@@ -318,6 +318,7 @@
     <ClCompile Include="DirectXTexNormalMaps.cpp" />
     <ClCompile Include="DirectXTexPMAlpha.cpp" />
     <ClCompile Include="DirectXTexResize.cpp" />
+    <ClCompile Include="DirectXTexSwizzle.cpp" />
```
We'll need to update all 9 vcxproj/filters sets to add this file. I can deal with that when you are ready.

walbourn commented on DirectXTex/DirectXTexSwizzle.cpp:
```diff
@@ -0,0 +1,303 @@
+    if (!srcImages || !depth || !IsValid(metadata.format) || depth > metadata.depth || !result.GetImages())
+        return E_INVALIDARG;
+
+    if (metadata.IsVolumemap()
```
For ``StandardSwizzle3D``, you want:

```
if (!metadata.IsVolumemap() || IsTypeless(metadata.format) || IsPlanar(metadata.format) || IsPalettized(metadata.format)))
    return E_INVALIDARG;
```

and for both you need to remove the ``IsCompressed`` test from this early out since you are adding compress handlnig.

isplunke replied:
duplicate of https://github.com/microsoft/DirectXTex/pull/515#discussion_r1765398888

isplunke committed:
Undo: Almost handel Block Compression.

isplunke committed:
name the file in itself

isplunke committed:
handle compression... this time?

isplunke committed:
use extractBits

walbourn commented on DirectXTex/DirectXTexSwizzle.cpp:
```diff
@@ -0,0 +1,285 @@
+//-------------------------------------------------------------------------------------
+// DirectXTexSwizzle.cpp
+//
+// DirectX Texture Library - Image container
```
Sorry, I didn't notice this the first time.

```
// DirectX Texture Library - Standard Swizzle (z-order curve)
```

walbourn commented on DirectXTex/DirectXTexSwizzle.cpp:
```diff
@@ -0,0 +1,285 @@
+_Use_decl_annotations_
+HRESULT DirectX::StandardSwizzle(const Image* srcImages, size_t nimages, const TexMetadata& metadata, bool toSwizzle, ScratchImage& result) noexcept
+{
+    HRESULT hr = result.Initialize2D(srcImages[0].format, srcImages[0].width, srcImages[0].height, nimages, 1);
```
Ideally for the error state coming out of all DirectXTex functions, the ``ScratchImage`` should be empty. I usually do all parameter validation first, then create the result working space, and at any failure-point after that I explicitly call ``result.Release();``

Also, you can let `ScratchImage` do a lot of the work for you setting up complex surfaces.

```
if (!srcImages || !nimages || !IsValid(metadata.format))
        return E_INVALIDARG;

if (metadata.IsVolumemap() || IsTypeless(metadata.format) || IsPlanar(metadata.format) || IsPalettized(metadata.format))
        return HRESULT_E_NOT_SUPPORTED;
        
HRESULT hr = result.Initialize(metadata);
if (FAILED(hr)
    return hr;
```

walbourn commented on DirectXTex/DirectXTexSwizzle.cpp:
```diff
@@ -0,0 +1,285 @@
+    if (metadata.IsVolumemap() || IsTypeless(metadata.format) || IsPlanar(metadata.format) || IsPalettized(metadata.format))
+        return HRESULT_E_NOT_SUPPORTED;
+
+    if (srcImages[0].format != metadata.format || srcImages[0].width != metadata.width || srcImages[0].height != metadata.height)
```
For `ScratchImage` you can guarantee that all the images are the same format since they are created that way by code we control, but for `Image` array input I validate that they are the same as the created result as part of the loop. Our only real concern here is making sure that the `metadata` passed properly describes the input array of images that we were given.

```
    for (size_t imageIndex = 0; imageIndex < nimages; ++imageIndex)
    {
        const Image& src = srcImages[index];
        if (src.format != metadata.format)
        {
            result.Release();
            return E_FAIL;
        }

        if ((src.width > UINT32_MAX) || (src.height > UINT32_MAX))
        {
            result.Release();
            return E_FAIL;
        }
       
        const uint8_t* sptr = src.pixels;
        if (!sptr)
        {
            result.Release();
            return E_POINTER;
        }

        const Image& dst = result.GetImages()[index];
        assert(dst.format == metadata.format);

        if (src.width != dst.width || src.height != dst.height)
        {
            result.Release();
            return E_FAIL;
        }

        uint8_t* dptr = dest.pixels;
        if (!dptr)
        {
            result.Release();
            return E_POINTER;
        }
```

walbourn commented on DirectXTex/DirectXTexSwizzle.cpp:
```diff
@@ -0,0 +1,285 @@
+    if (!dptr)
+        return E_POINTER;
+
+    size_t bytesPerPixel = BitsPerPixel(srcImage.format) / 8;
```
With the new function in main you can merge in:

```
const bool isCompressed = IsCompressed(srcImage.format);
size_t bytesPerPixel = (isCompressed() BytesPerBlock(srcImage.format) ? (BitsPerPixel(srcImage.format) / 8);
```


```
size_t height = isCompressed ? std::max(1, srcImage.height + 3) / 4) : srcImage.height;
size_t width  = isCompressed ? std::max(1, srcImage.width  + 3) / 4 ) : srcImage.width;
```

walbourn commented on DirectXTex/DirectXTexSwizzle.cpp:
```diff
@@ -0,0 +1,285 @@
+                uint32_t swizzleIndex = deposit_bits(x, xBytesMask) + deposit_bits(y, yBytesMask);
+                size_t swizzleOffset = swizzleIndex * bytesPerPixel;
+
+                size_t rowMajorOffset = y * rowPitch + x * bytesPerPixel;
```
rowMajor is efficiently handled with:

```
for (size_t y = 0; y < height; y++)
{
    const uint8_t* sourcePixelPointer = sptr;
    for (size_t x = 0; x < width; x++)
    {
        uint32_t swizzleIndex = deposit_bits(x, xBytesMask) + deposit_bits(y, yBytesMask);
        size_t swizzleOffset = swizzleIndex * bytesPerPixel;
        uint8_t* destPixelPointer = dptr + swizzleOffset;
        memcpy(destPixelPointer, sourcePixelPointer, bytesPerPixel);
        sourcePixelPointer += bytesPerPixel;
    }
    sptr += rowPitch;
}
                
```

walbourn committed:
Update DirectXTexSwizzle.cpp

walbourn committed:
Update DirectXTexSwizzle.cpp

More code review

walbourn committed:
Update DirectXTexSwizzle.cpp

Rework 2D swizzle into templates

walbourn committed:
Update DirectXTexSwizzle.cpp

Code review

walbourn committed:
Update CMakeLists.txt

Linux filesystem is case sensitive

walbourn committed:
Update remaining vcxproj for new file

walbourn committed:
StandardSwizzle3D merged into main function

walbourn committed:
More work on 3D swizzle

walbourn committed:
More code review

walbourn committed:
Attempt to fix CodeQL warning

walbourn committed:
Add 3D case for linear to swizzle

walbourn committed:
Add DIRECTX_TEX_API to StandardSwizzle functions

Updated with DLL annotation for new function

walbourn committed:
Implement 3D swizzle to linear

fdwr commented on DirectXTex/DirectXTexSwizzle.cpp:
```diff
@@ -0,0 +1,755 @@
+    uint32_t extract_bits(uint32_t val, int mask) noexcept
+    {
+        uint32_t res = 0;
+        for (uint32_t bb = 1; mask !=0; bb += bb)
```
Also, unless there's a perf advantage to left shifting via addition (maybe there is), just saying `bb <<= 1` would probably be clearer to readers.

fdwr commented on DirectXTex/DirectXTexSwizzle.cpp:
```diff
@@ -0,0 +1,755 @@
+    constexpr uint64_t MAX_TEXTURE_SIZE = UINT32_MAX;
+#endif
+
+    // Standard Swizzle is not defined for these formats.
```
Q: I'm curious where this list came from, from documentation or trial-and-error calling the API? Is `DXGI_FORMAT_R8_SNORM` valid, given it's not in the `switch`? I see mention in [the documentation](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/ne-d3d12-d3d12_texture_layout) that `DXGI_FORMAT_R32G32B32_TYPELESS` is excluded (along with MSAA, 1D, depth and stencil), but otherwise I'm not seeing any particular list.

walbourn replied:
I believe I made this list based on a spec. This function is very much specific to `64KB_STANDARD_SWIZZLE`.

fdwr reviewed (commented):
👀

walbourn committed:
Delete DirectXTex/DirectXTex_Desktop_2019.vcxproj

walbourn committed:
Delete DirectXTex/DirectXTex_Desktop_2019.vcxproj.filters

walbourn committed:
Delete DirectXTex/DirectXTex_Desktop_2019_Win10.vcxproj

walbourn committed:
Delete DirectXTex/DirectXTex_Desktop_2019_Win10.vcxproj.filters

walbourn committed:
Delete DirectXTex/DirectXTex_GDK_2019.vcxproj

walbourn committed:
Delete DirectXTex/DirectXTex_GDK_2019.vcxproj.filters

walbourn committed:
Delete DirectXTex/DirectXTex_GXDK_PC_2019.vcxproj

walbourn committed:
Delete DirectXTex/DirectXTex_GXDK_PC_2019.vcxproj.filters

walbourn committed:
Revert all vcxproj changes

walbourn commented on CMakeLists.txt:
```diff
@@ -97,6 +97,7 @@ set(LIBRARY_SOURCES
     DirectXTex/DirectXTexMipmaps.cpp
     DirectXTex/DirectXTexMisc.cpp
     DirectXTex/DirectXTexNormalMaps.cpp
+    DirectXTex/DirectXTexSwizzle.cpp
```
I've reverted all the vcxproj changes for now. I'll do those in a follow-up PR after finishing this one.

walbourn committed:
Code review feedback

walbourn committed:
Update DirectXTexSwizzle.cpp

dep/extract was inversed. Fixed for 2D

fdwr commented on DirectXTex/DirectXTexSwizzle.cpp:
```diff
@@ -206,43 +227,46 @@ namespace
-        const uint8_t* endPtr = sptr + static_cast<ptrdiff_t>(totalDataSize);
-        for (size_t swizzleIndex = 0; swizzleIndex < static_cast<size_t>(totalPixels); ++swizzleIndex)
+        const uint8_t* endPtr = sptr + tileSizeInBytes;
+        for(uint32_t yy = 0; yy < height; ++yy)
```
(nit style consistency)
```suggestion
       for (uint32_t yy = 0; yy < height; ++yy)
```

Below too
```c++
for(uint32_t xx = 0; xx < widthInBytes; xx += bytesPerPixelOrBlock)
```

fdwr commented on DirectXTex/DirectXTexSwizzle.cpp:
```diff
@@ -132,60 +134,79 @@ namespace
+                // This path pads with copies of the final block rather than zero-fill.
+                if (xx < widthInBytes && yy < height)
+                {
+                    const uint8_t * sourcePixelPointer = sptr + (yy * rowPitch) + xx;
```
(nit, below too)
```suggestion
                   const uint8_t* sourcePixelPointer = sptr + (yy * rowPitch) + xx;
```

fdwr commented on DirectXTex/DirectXTexSwizzle.cpp:
```diff
@@ -702,23 +744,38 @@ HRESULT DirectX::StandardSwizzle(
+                        break;
+                    case 2:
+                        hr = LinearToStandardSwizzle2D<STANDARD_SWIZZLE_MASK_16, 2, false>(src, dst);
+                        break;
```
I think an explicit `case 4` would be clearer rather than falling through `default` (because if you ever changed the default case later, it would impact 4 case as collateral). The compiler can merge the logical blocks.

fdwr reviewed (commented):
👀
